File: code/Mines.Abstractions/IMineGrid.cs
namespace Mikodev.Mines.Abstractions
{
    /// <summary>
    /// 雷区
    /// </summary>
    public interface IMineGrid
    {
        /// <summary>
        /// 游戏结束
        /// </summary>
        bool IsGameOver { get; }

        /// <summary>
        /// X 最大值
        /// </summary>
        int XMax { get; }

        /// <summary>
        /// Y 最大值
        /// </summary>
        int YMax { get; }

        /// <summary>
        /// 获取指定坐标的内容
        /// </summary>
        MineMark Get(int x, int y);

        /// <summary>
        /// 设置标记 (旗帜, 问号, 或清除标记)
        /// </summary>
        void Set(int x, int y);

        /// <summary>
        /// 翻开方块
        /// </summary>
        void Remove(int x, int y);

        /// <summary>
        /// 根据旗帜翻开周围方块
        /// </summary>
        void RemoveAll(int x, int y);
    }
}

File: code/Mines.Annotations/IMineGrid.fs
namespace Mikodev.Mines.Annotations

open System

[<AllowNullLiteral>]
type IMineGrid =
    [<CLIEvent>]
    abstract member StatusChanged : IEvent<EventHandler, EventArgs>

    abstract member Status : MineGridStatus

    abstract member XMax : int

    abstract member YMax : int

    abstract member FlagCount : int

    abstract member MineCount : int

    abstract member Get : x : int * y : int -> MineData

    abstract member Set : x : int * y : int * mark : MineMark -> unit

    abstract member Remove : x : int * y : int -> int
